### PR TITLE
Automatic backend discovery support based on Cluster API

### PR DESCRIPTION
**Pull request by nurali-techie**
### What does this PR do?
Introduce new custom traefik backend named 'OSIO' which call Cluster API periodically and re-configure traefik backends. Here is link to one of the existing 'file' backend provided by traefik - https://docs.traefik.io/configuration/backends/file/

Fix - https://github.com/fabric8-services/fabric8-oso-proxy/issues/3 under https://github.com/openshiftio/openshift.io/issues/1988

### Motivation
As and when new OS cluster is introduced, it will be automatically configured with traefik as backend.

### More
One integration test added to cover end-to-end OSIO provider functionality.

### CODE REVIEW

File: configuration/configuration.go [hunk 1]
```diff
@@ -257,6 +260,20 @@ func (gc *GlobalConfiguration) SetEffectiveConfiguration(configFile string) {
...
+			panic("Missing SERVICE_ACCOUNT_ID")
+		}
+		gc.OSIO.ServiceAccountID = saID
+
+		saSecret := os.Getenv("SERVICE_ACCOUNT_SECRET")
+		if len(saSecret) <= 0 {
+			panic("Missing SERVICE_ACCOUNT_SECRET")
+		}
+		gc.OSIO.ServiceAccountSecret = saSecret
+	}
```
Comment (nurali-techie) [inline configuration/configuration.go]: get security param from environment and set back to OSIO provider.

File: configuration/configuration.go [hunk 2]
```diff
@@ -94,6 +96,7 @@ type GlobalConfiguration struct {
 	API                       *api.Handler            `description:"Enable api/dashboard" export:"true"`
 	Metrics                   *types.Metrics          `description:"Enable a metrics exporter" export:"true"`
 	Ping                      *ping.Handler           `description:"Enable ping" export:"true"`
+	OSIO                      *osio.Provider          `description:"Enable OSIO backend with default settings" export:"true"`
```
Comment (nurali-techie) [inline configuration/configuration.go]: This will hold OSIO provider with traefik configuration.

File: integration/osio_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	// Start OSIO servers
+	ts1 := startOSIOServer(8081, nil)
+	defer ts1.Close()
+	ts2 := startOSIOServer(8082, nil)
+	defer ts2.Close()
+
+	// make multiple reqeust on some time gap
+	// note, req has 'Bearer 2222' so it should go to 'http://127.0.0.1:8082' check serverAUTHRequest2()
+	// check serverAUTHRequest2(), return 'http://127.0.0.1:8082' cluster for only first time
+	// so first few response would be 'HTTP 200 OK' and then rest would be 'HTTP 404 not found'
```
Comment (nurali-techie) [inline integration/osio_provider_test.go]: Check comments in test case.  This test cover end-to-end OSIO backend functionality.

File: provider/osio/osio.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	defaultBackendURL string
+}
+
+// Provide allows the osio provider to provide configurations to traefik
+// using the given configuration channel.
+func (p *Provider) Provide(configChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
+	log.Debugf("Configuring %s provider", providerName)
+	p.init(configChan)
+
+	handleCanceled := func(ctx context.Context, err error) error {
```
Comment (nurali-techie) [inline provider/osio/osio.go]: Standard way in traefik to configure "wait -> pull" for periodically update configuration.  Please check - https://github.com/fabric8-services/fabric8-oso-proxy/blob/master/provider/ecs/ecs.go#L108

This has backoff and stop by traefik support.

File: provider/osio/osio.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+		}
+		configInd := fmt.Sprintf("%d", ind+1)
+		config.Frontends["frontend"+configInd] = createFrontend(cluster.APIURL, "backend"+configInd)
+		config.Backends["backend"+configInd] = createBackend(cluster.APIURL)
+	}
+	if !defaultBackendExist {
+		p.defaultBackendURL = clusterResp.Clusters[0].APIURL
+		config.Frontends["default"] = createFrontend("default", "default")
+		config.Backends["default"] = createBackend(p.defaultBackendURL)
+	}
```
Comment (nurali-techie) [inline provider/osio/osio.go]: Add default provider to support https://github.com/fabric8-services/fabric8-oso-proxy/blob/master/middlewares/osio/auth.go#L94

File: server/server.go [hunk 6]
```diff
@@ -586,7 +586,9 @@ func (s *Server) configureProviders() {
 	if s.globalConfiguration.ServiceFabric != nil {
 		s.providers = append(s.providers, s.globalConfiguration.ServiceFabric)
 	}
-
+	if s.globalConfiguration.OSIO != nil {
+		s.providers = append(s.providers, s.globalConfiguration.OSIO)
```
Comment (nurali-techie) [inline server/server.go]: Register/Add OSIO traefik backend with traefik.

File: openshift/f8osoproxy.secrets.yml [hunk 7]
```diff
@@ -12,4 +12,7 @@ objects:
   data:
     wit.url: https://api.prod-preview.openshift.io/api
     auth.url: https://auth.prod-preview.openshift.io/api
-    traefik.toml: ""
\ No newline at end of file
+    traefik.toml: ""
+    service.account.id: YzIxMWYxYmQtMTdhNy00ZjhjLTlmODAtMDkxN2QxNjc4ODlk
```
Comment (aslakknutsen) [inline openshift/f8osoproxy.secrets.yml]: Move these to a Secret. Not in the ConfigMap

File: integration/fixtures/osio_provider_config.toml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+checkNewVersion = false
+
+[api]
+entryPoint = "traefik"
+
+[osio]
+refreshSeconds = 3
+clusterAPI = "http://127.0.0.1:9091/api/clusters"
+tokenAPI = "http://127.0.0.1:9091/api/token"
```
Comment (nurali-techie) [inline integration/fixtures/osio_provider_config.toml]: @aslakknutsen This is configuration for 'osio' traefik backend.  Please add above variable in traefik.toml file in f8-oso-proxy project.  As decided, refreshSeconds = 900 (i.e. 15 mins) for production.

**Comment by nurali-techie** [comment]
@aslakknutsen checking test failure in jenkin.

**Comment by nurali-techie** [comment]
@surajssd Please have a look and provide your feedback.

File: provider/osio/osio.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+// Provide allows the osio provider to provide configurations to traefik
+// using the given configuration channel.
+func (p *Provider) Provide(configChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
+	log.Debugf("Configuring %s provider", providerName)
+	p.init(configChan)
+	p.scheduleConfigPull(configChan, pool)
+	return nil
+}
+
+func (p *Provider) scheduleConfigPull(configChan chan<- types.ConfigMessage, pool *safe.Pool) {
```
Comment (surajssd) [inline provider/osio/osio.go]: is this `pull` or `poll` or `pool` I am confused here, also a comment would be helpful?

File: provider/osio/client.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("Call to token api failed, code:%d, error:%s", resp.StatusCode, resp.Status)
+	}
+
+	defer resp.Body.Close()
+	var tokenResp *tokenResponse
+	json.NewDecoder(resp.Body).Decode(&tokenResp)
```
[the remark this answers is not in the document]
Comment (nurali-techie) [reply provider/osio/client.go]: @aslakknutsen thanks for this.

File: provider/osio/client.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	Name       string `json:"name"`
+}
+
+type clusterResponse struct {
+	Clusters []clusterData `json:"data"`
+}
+
+func (client *authClient) callTokenAPI(tokenAPI string, tokenReq *tokenRequest) (*tokenResponse, error) {
+	reqBody := new(bytes.Buffer)
+	json.NewEncoder(reqBody).Encode(tokenReq)
```
[the remark this answers is not in the document]
Comment (nurali-techie) [reply provider/osio/client.go]: @aslakknutsen thanks for fixing this.

File: provider/osio/osio.go (hunk 9 above)
Comment (nurali-techie) [reply provider/osio/osio.go]: now it's changed to schedule()

**Comment by nurali-techie** [comment]
@surajssd incorporated all your comment actually by @aslakknutsen (thanks :-)).
Can you please do review ok.

**Pull request merged**